crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -1908,14 +1911,11 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> BlockWriter for DatabaseProvider<'
         let tx_iter = if Some(block.body.len()) == senders_len {
             block.body.into_iter().zip(senders.unwrap()).collect::<Vec<(_, _)>>()
         } else {
-            block
-                .body
-                .into_iter()
-                .map(|tx| {
-                    let signer = tx.recover_signer();
-                    (tx, signer.unwrap_or_default())
-                })
-                .collect::<Vec<(_, _)>>()
+            let senders = TransactionSigned::recover_signers(&block.body, block.body.len()).ok_or(
+                BlockExecutionError::Validation(BlockValidationError::SenderRecoveryError),
+            )?;
+
+            block.body.into_iter().zip(senders).collect()
```

Comment:
could you pls add `debug_assert!(senders.len(), block.body.len())` here? wanna be extra cautious since using `zip` bit us couple of times